File: hiob_contracts/storymap_v1.py
"""Strict, immutable Story OS maps and bounded experiment treatments.

Python is authoritative. A StoryMap holds customer and evidence truth; one
hypothesis and every treatment bind to its digest. Treatments deliberately have
no identity, product, or proof-fact fields: only hook, proof order, framing,
and CTA may vary.
"""

from __future__ import annotations

from typing import Annotated, Any, Literal, Mapping

from pydantic import BaseModel, Field, field_validator, model_validator

from .ares_script_revision_v1 import (
    DigestStr,
    NonBlankStr,
    _FROZEN_STRICT,
    canonical_contract_digest_v1,
)


MAX_PROOF_REFERENCES_V1 = 12
MAX_VARIANTS_V1 = 12

_STORY_MAP_DIGEST_FIELDS = (
    "contract_version",
    "customer_scene",
    "bad_alternative_tension",
    "urgent_moment",
    "emotional_stake",
    "proof_references",
    "objection",
    "offer",
    "cta",
    "target_metric",
    "content_mode",
    "story_policy_digest",
)
_EXPERIMENT_HYPOTHESIS_DIGEST_FIELDS = (
    "contract_version",
    "story_map_digest",
    "hypothesis",
)
_VARIANT_SET_DIGEST_FIELDS = (
    "contract_version",
    "story_map",
    "story_map_digest",
    "experiment_hypothesis",
    "variants",
)


def _data(value: Mapping[str, Any] | BaseModel) -> Mapping[str, Any]:
    return value.model_dump(mode="json") if isinstance(value, BaseModel) else value


def _payload(
    value: Mapping[str, Any] | BaseModel,
    fields: tuple[str, ...],
) -> dict[str, Any]:
    data = _data(value)
    return {field: data[field] for field in fields}


def derive_story_map_digest_v1(value: Mapping[str, Any] | BaseModel) -> str:
    """Derive the content digest for one immutable customer story map."""

    return canonical_contract_digest_v1(_payload(value, _STORY_MAP_DIGEST_FIELDS))


def derive_experiment_hypothesis_digest_v1(
    value: Mapping[str, Any] | BaseModel,
) -> str:
    """Derive a hypothesis digest bound to exactly one StoryMap digest."""

    return canonical_contract_digest_v1(
        _payload(value, _EXPERIMENT_HYPOTHESIS_DIGEST_FIELDS)
    )


def derive_variant_set_digest_v1(value: Mapping[str, Any] | BaseModel) -> str:
    """Derive the complete set digest, including every bounded treatment."""

    return canonical_contract_digest_v1(_payload(value, _VARIANT_SET_DIGEST_FIELDS))


class ProofReferenceV1(BaseModel):
    """A stable reference ID and its locked proof-fact digest."""

    model_config = _FROZEN_STRICT

    proof_ref_id: Annotated[NonBlankStr, Field(max_length=160)]
    proof_fact_digest: DigestStr


class StoryMapV1(BaseModel):
    """The complete customer-and-evidence truth for a Story OS experiment."""

    model_config = _FROZEN_STRICT

    contract_version: Literal["StoryMap.v1"]
    customer_scene: Annotated[NonBlankStr, Field(max_length=1_200)]
    bad_alternative_tension: Annotated[NonBlankStr, Field(max_length=1_200)]
    urgent_moment: Annotated[NonBlankStr, Field(max_length=600)]
    emotional_stake: Annotated[NonBlankStr, Field(max_length=600)]
    proof_references: tuple[ProofReferenceV1, ...] = Field(
        min_length=1,
        max_length=MAX_PROOF_REFERENCES_V1,
    )
    objection: Annotated[NonBlankStr, Field(max_length=600)]
    offer: Annotated[NonBlankStr, Field(max_length=600)]
    cta: Annotated[NonBlankStr, Field(max_length=300)]
    target_metric: Annotated[NonBlankStr, Field(max_length=160)]
    content_mode: Literal["ugc", "information"]
    story_policy_digest: DigestStr
    story_map_digest: DigestStr

    @field_validator("proof_references", mode="before")
    @classmethod
    def _proof_references_are_immutable_tuple(cls, value: Any) -> Any:
        return tuple(value) if isinstance(value, list) else value

    @model_validator(mode="after")
    def _story_map_is_complete_and_bound(self) -> "StoryMapV1":
        proof_ref_ids = [reference.proof_ref_id for reference in self.proof_references]
        if len(set(proof_ref_ids)) != len(proof_ref_ids):
            raise ValueError("proof_references contains duplicate proof_ref_id")
        if self.story_map_digest != derive_story_map_digest_v1(self):
            raise ValueError("story_map_digest does not match StoryMap content")
        return self


class ExperimentHypothesisV1(BaseModel):
    """One measurable claim about a StoryMap; no truth fields are copied here."""

    model_config = _FROZEN_STRICT

    contract_version: Literal["ExperimentHypothesis.v1"]
    story_map_digest: DigestStr
    hypothesis: Annotated[NonBlankStr, Field(max_length=1_200)]
    experiment_hypothesis_digest: DigestStr

    @model_validator(mode="after")
    def _hypothesis_is_bound(self) -> "ExperimentHypothesisV1":
        if self.experiment_hypothesis_digest != derive_experiment_hypothesis_digest_v1(
            self
        ):
            raise ValueError(
                "experiment_hypothesis_digest does not match ExperimentHypothesis content"
            )
        return self


class _StoryVariantV1(BaseModel):
    """A treatment deliberately restricted to the four allowed axes."""

    model_config = _FROZEN_STRICT

    variant_id: Annotated[NonBlankStr, Field(max_length=160)]
    story_map_digest: DigestStr
    hook: Annotated[NonBlankStr, Field(max_length=600)]
    proof_order: tuple[Annotated[NonBlankStr, Field(max_length=160)], ...] = Field(
        min_length=1,
        max_length=MAX_PROOF_REFERENCES_V1,
    )
    framing: Annotated[NonBlankStr, Field(max_length=1_200)]
    cta: Annotated[NonBlankStr, Field(max_length=300)]

    @field_validator("proof_order", mode="before")
    @classmethod
    def _proof_order_is_immutable_tuple(cls, value: Any) -> Any:
        return tuple(value) if isinstance(value, list) else value


class VariantSetV1(BaseModel):
    """A StoryMap-bound set of treatments with no fact-drift surface."""

    model_config = _FROZEN_STRICT

    contract_version: Literal["VariantSet.v1"]
    story_map: StoryMapV1
    story_map_digest: DigestStr
    experiment_hypothesis: ExperimentHypothesisV1
    variants: tuple[_StoryVariantV1, ...] = Field(
        min_length=1,
        max_length=MAX_VARIANTS_V1,
    )
    variant_set_digest: DigestStr

    @field_validator("variants", mode="before")
    @classmethod
    def _variants_are_immutable_tuple(cls, value: Any) -> Any:
        return tuple(value) if isinstance(value, list) else value

    @model_validator(mode="after")
    def _variant_set_is_story_map_bound(self) -> "VariantSetV1":
        if self.story_map_digest != self.story_map.story_map_digest:
            raise ValueError("story_map_digest does not match embedded StoryMap")
        if (
            self.experiment_hypothesis.story_map_digest
            != self.story_map.story_map_digest
        ):
            raise ValueError("experiment_hypothesis is bound to a different story_map_digest")

        variant_ids = [variant.variant_id for variant in self.variants]
        if len(set(variant_ids)) != len(variant_ids):
            raise ValueError("variants contains duplicate variant_id")

        expected_proof_ids = tuple(
            reference.proof_ref_id for reference in self.story_map.proof_references
        )
        for variant in self.variants:
            if variant.story_map_digest != self.story_map.story_map_digest:
                raise ValueError("variant story_map_digest does not match VariantSet")
            if (
                len(variant.proof_order) != len(expected_proof_ids)
                or set(variant.proof_order) != set(expected_proof_ids)
            ):
                raise ValueError(
                    "variant proof_order must be an exact ordering of StoryMap proof_ref_id values"
                )

        if self.variant_set_digest != derive_variant_set_digest_v1(self):
            raise ValueError("variant_set_digest does not match VariantSet content")
        return self


__all__ = [
    "ExperimentHypothesisV1",
    "MAX_PROOF_REFERENCES_V1",
    "MAX_VARIANTS_V1",
    "ProofReferenceV1",
    "StoryMapV1",
    "VariantSetV1",
    "derive_experiment_hypothesis_digest_v1",
    "derive_story_map_digest_v1",
    "derive_variant_set_digest_v1",
]

File: ts/src/storymap-v1.ts
/** Strict Story OS maps and the only four permitted experiment treatment axes. */
import { z } from 'zod';

import { sha256Digest } from './factory/digest.js';

export const MAX_PROOF_REFERENCES_V1 = 12;
export const MAX_VARIANTS_V1 = 12;

const DigestSchema = z
  .string()
  .regex(/^sha256:[0-9a-f]{64}$/, 'digest must be sha256:<64 lowercase hex>');

function hasUnpairedSurrogate(value: string): boolean {
  for (let index = 0; index < value.length; index += 1) {
    const code = value.charCodeAt(index);
    if (code >= 0xd800 && code <= 0xdbff) {
      const low = value.charCodeAt(index + 1);
      if (!(low >= 0xdc00 && low <= 0xdfff)) return true;
      index += 1;
    } else if (code >= 0xdc00 && code <= 0xdfff) {
      return true;
    }
  }
  return false;
}

function assertJsonUnicodeScalars(value: unknown): void {
  if (typeof value === 'string') {
    if (hasUnpairedSurrogate(value)) {
      throw new TypeError('text must contain valid Unicode scalar values');
    }
    return;
  }
  if (Array.isArray(value)) {
    for (const item of value) assertJsonUnicodeScalars(item);
    return;
  }
  if (value !== null && typeof value === 'object') {
    for (const [key, item] of Object.entries(value as Record<string, unknown>)) {
      if (hasUnpairedSurrogate(key)) {
        throw new TypeError('text must contain valid Unicode scalar values');
      }
      assertJsonUnicodeScalars(item);
    }
  }
}

function isPythonStripWhitespace(char: string): boolean {
  const code = char.codePointAt(0);
  return code !== undefined && (
    (code >= 0x0009 && code <= 0x000d)
    || (code >= 0x001c && code <= 0x001f)
    || code === 0x0020
    || code === 0x0085
    || code === 0x00a0
    || code === 0x1680
    || (code >= 0x2000 && code <= 0x200a)
    || code === 0x2028
    || code === 0x2029
    || code === 0x202f
    || code === 0x205f
    || code === 0x3000
  );
}

function isPythonBlank(value: string): boolean {
  return value.length === 0 || [...value].every(isPythonStripWhitespace);
}

const boundedNonBlankString = (maxLength: number) => z
  .string()
  .refine((value) => !isPythonBlank(value), 'string must not be blank')
  .refine(
    (value) => !hasUnpairedSurrogate(value),
    'text must contain valid Unicode scalar values',
  )
  .refine(
    (value) => [...value].length <= maxLength,
    `string must contain at most ${maxLength} Unicode scalars`,
  );

function deepFreeze<T>(value: T): T {
  if (value !== null && typeof value === 'object') {
    for (const child of Object.values(value as Record<string, unknown>)) {
      deepFreeze(child);
    }
    Object.freeze(value);
  }
  return value;
}

function requiredField(value: Record<string, unknown>, field: string): unknown {
  if (!(field in value) || value[field] === undefined) {
    throw new TypeError(`${field} is required for Story OS digest`);
  }
  return value[field];
}

function digestPayload(
  value: Record<string, unknown>,
  fields: readonly string[],
): Record<string, unknown> {
  return Object.fromEntries(
    fields.map((field) => [field, requiredField(value, field)]),
  );
}

export function deriveStoryMapDigestV1(value: Record<string, unknown>): string {
  const payload = digestPayload(value, [
    'contract_version', 'customer_scene', 'bad_alternative_tension',
    'urgent_moment', 'emotional_stake', 'proof_references', 'objection',
    'offer', 'cta', 'target_metric', 'content_mode', 'story_policy_digest',
  ]);
  assertJsonUnicodeScalars(payload);
  return sha256Digest(payload);
}

export function deriveExperimentHypothesisDigestV1(
  value: Record<string, unknown>,
): string {
  const payload = digestPayload(value, [
    'contract_version', 'story_map_digest', 'hypothesis',
  ]);
  assertJsonUnicodeScalars(payload);
  return sha256Digest(payload);
}

export function deriveVariantSetDigestV1(value: Record<string, unknown>): string {
  const payload = digestPayload(value, [
    'contract_version', 'story_map', 'story_map_digest',
    'experiment_hypothesis', 'variants',
  ]);
  assertJsonUnicodeScalars(payload);
  return sha256Digest(payload);
}

const ProofReferenceV1Schema = z.object({
  proof_ref_id: boundedNonBlankString(160),
  proof_fact_digest: DigestSchema,
}).strict();

export const StoryMapV1Schema = z.object({
  contract_version: z.literal('StoryMap.v1'),
  customer_scene: boundedNonBlankString(1200),
  bad_alternative_tension: boundedNonBlankString(1200),
  urgent_moment: boundedNonBlankString(600),
  emotional_stake: boundedNonBlankString(600),
  proof_references: z.array(ProofReferenceV1Schema).min(1).max(MAX_PROOF_REFERENCES_V1),
  objection: boundedNonBlankString(600),
  offer: boundedNonBlankString(600),
  cta: boundedNonBlankString(300),
  target_metric: boundedNonBlankString(160),
  content_mode: z.enum(['ugc', 'information']),
  story_policy_digest: DigestSchema,
  story_map_digest: DigestSchema,
}).strict().superRefine((value, ctx) => {
  const ids = value.proof_references.map((reference) => reference.proof_ref_id);
  if (new Set(ids).size !== ids.length) {
    ctx.addIssue({
      code: z.ZodIssueCode.custom,
      path: ['proof_references'],
      message: 'proof_references contains duplicate proof_ref_id',
    });
  }
  try {
    if (value.story_map_digest === deriveStoryMapDigestV1(value)) return;
    ctx.addIssue({
      code: z.ZodIssueCode.custom,
      path: ['story_map_digest'],
      message: 'story_map_digest does not match StoryMap content',
    });
  } catch {
    ctx.addIssue({
      code: z.ZodIssueCode.custom,
      path: [],
      message: 'StoryMap text must contain valid Unicode scalar values',
    });
  }
}).transform(deepFreeze);

export type StoryMapV1 = z.infer<typeof StoryMapV1Schema>;

export const ExperimentHypothesisV1Schema = z.object({
  contract_version: z.literal('ExperimentHypothesis.v1'),
  story_map_digest: DigestSchema,
  hypothesis: boundedNonBlankString(1200),
  experiment_hypothesis_digest: DigestSchema,
}).strict().superRefine((value, ctx) => {
  try {
    if (value.experiment_hypothesis_digest === deriveExperimentHypothesisDigestV1(value)) return;
    ctx.addIssue({
      code: z.ZodIssueCode.custom,
      path: ['experiment_hypothesis_digest'],
      message: 'experiment_hypothesis_digest does not match ExperimentHypothesis content',
    });
  } catch {
    ctx.addIssue({
      code: z.ZodIssueCode.custom,
      path: [],
      message: 'ExperimentHypothesis text must contain valid Unicode scalar values',
    });
  }
}).transform(deepFreeze);

export type ExperimentHypothesisV1 = z.infer<typeof ExperimentHypothesisV1Schema>;

const StoryVariantV1Schema = z.object({
  variant_id: boundedNonBlankString(160),
  story_map_digest: DigestSchema,
  hook: boundedNonBlankString(600),
  proof_order: z.array(boundedNonBlankString(160)).min(1).max(MAX_PROOF_REFERENCES_V1),
  framing: boundedNonBlankString(1200),
  cta: boundedNonBlankString(300),
}).strict();

export const VariantSetV1Schema = z.object({
  contract_version: z.literal('VariantSet.v1'),
  story_map: StoryMapV1Schema,
  story_map_digest: DigestSchema,
  experiment_hypothesis: ExperimentHypothesisV1Schema,
  variants: z.array(StoryVariantV1Schema).min(1).max(MAX_VARIANTS_V1),
  variant_set_digest: DigestSchema,
}).strict().superRefine((value, ctx) => {
  if (value.story_map_digest !== value.story_map.story_map_digest) {
    ctx.addIssue({
      code: z.ZodIssueCode.custom,
      path: ['story_map_digest'],
      message: 'story_map_digest does not match embedded StoryMap',
    });
  }
  if (value.experiment_hypothesis.story_map_digest !== value.story_map.story_map_digest) {
    ctx.addIssue({
      code: z.ZodIssueCode.custom,
      path: ['experiment_hypothesis', 'story_map_digest'],
      message: 'experiment_hypothesis is bound to a different story_map_digest',
    });
  }
  const variantIds = value.variants.map((variant) => variant.variant_id);
  if (new Set(variantIds).size !== variantIds.length) {
    ctx.addIssue({
      code: z.ZodIssueCode.custom,
      path: ['variants'],
      message: 'variants contains duplicate variant_id',
    });
  }
  const expectedProofIds = value.story_map.proof_references.map(
    (reference) => reference.proof_ref_id,
  );
  for (const [index, variant] of value.variants.entries()) {
    if (variant.story_map_digest !== value.story_map.story_map_digest) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        path: ['variants', index, 'story_map_digest'],
        message: 'variant story_map_digest does not match VariantSet',
      });
    }
    if (
      variant.proof_order.length !== expectedProofIds.length
      || new Set(variant.proof_order).size !== expectedProofIds.length
      || variant.proof_order.some((proofRefId) => !expectedProofIds.includes(proofRefId))
    ) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        path: ['variants', index, 'proof_order'],
        message: 'variant proof_order must be an exact ordering of StoryMap proof_ref_id values',
      });
    }
  }
  try {
    if (value.variant_set_digest === deriveVariantSetDigestV1(value)) return;
    ctx.addIssue({
      code: z.ZodIssueCode.custom,
      path: ['variant_set_digest'],
      message: 'variant_set_digest does not match VariantSet content',
    });
  } catch {
    ctx.addIssue({
      code: z.ZodIssueCode.custom,
      path: [],
      message: 'VariantSet text must contain valid Unicode scalar values',
    });
  }
}).transform(deepFreeze);

export type VariantSetV1 = z.infer<typeof VariantSetV1Schema>;
